firedrake/preconditioners/fdm.py

```diff
@@ -136,73 +187,960 @@ def interp_nullspace(I, nsp):
+        :arg J: the Jacobian bilinear form
+        :arg bcs: an iterable of boundary conditions on V
+        :arg form_compiler_parameters: parameters to assemble diagonal factors
+        :arg pmat_type: the preconditioner `PETSc.Mat.Type`
+        :arg use_static_condensation: are we assembling the statically-condensed Schur complement on facets?
+
+        :returns: 2-tuple with the preconditioner :class:`PETSc.Mat` and its assembly callable
+        """
+        ifacet, = numpy.nonzero([is_restricted(Vsub.finat_element)[1] for Vsub in V])
+        if len(ifacet) == 0:
+            Vfacet = None
+            Vbig = V
+            _, fdofs = split_dofs(V.finat_element)
+        elif len(ifacet) == 1:
+            Vfacet = V[ifacet[0]]
+            ebig, = set(unrestrict_element(Vsub.ufl_element()) for Vsub in V)
+            Vbig = firedrake.FunctionSpace(V.mesh(), ebig)
+            if len(V) > 1:
+                dims = [Vsub.finat_element.space_dimension() for Vsub in V]
+                assert sum(dims) == Vbig.finat_element.space_dimension()
+            fdofs = restricted_dofs(Vfacet.finat_element, Vbig.finat_element)
+        else:
+            raise ValueError("Expecting at most one FunctionSpace restricted onto facets.")
+
+        value_size = Vbig.value_size
+        if value_size != 1:
+            fdofs = numpy.add.outer(value_size * fdofs, numpy.arange(value_size, dtype=fdofs.dtype))
+        dofs = numpy.arange(value_size * Vbig.finat_element.space_dimension(), dtype=fdofs.dtype)
+        idofs = numpy.setdiff1d(dofs, fdofs, assume_unique=True)
+        self.ises = tuple(PETSc.IS().createGeneral(indices, comm=PETSc.COMM_SELF) for indices in (idofs, fdofs))
+        self.submats = [None for _ in range(7)]
+
+        self.reference_tensor_on_diag = dict()
+        self.get_static_condensation = dict()
+        if Vfacet and use_static_condensation:
+            # If we are in a facet space, we build the Schur complement on its diagonal block
+            self.reference_tensor_on_diag[Vfacet] = self.assemble_reference_tensor(Vbig)
+            self.get_static_condensation[Vfacet] = lambda A: condense_element_mat(A, self.ises[0], self.ises[1], self.submats)
+
+        elif len(fdofs) and V.finat_element.formdegree == 0:
+            # If we are in H(grad), we just pad with zeros on the statically-condensed pattern
+            i1 = PETSc.IS().createGeneral(dofs, comm=PETSc.COMM_SELF)
+            self.get_static_condensation[V] = lambda Ae: condense_element_pattern(Ae, self.ises[0], i1, self.submats)
+
+        @PETSc.Log.EventDecorator("FDMGetIndices")
+        def cell_to_global(lgmap, cell_to_local, cell_index, result=None):
+            # Be careful not to create new arrays
+            result = cell_to_local(cell_index, result=result)
+            return lgmap.apply(result, result=result)
+
+        # Create data structures needed for assembly
+        self.cell_to_global = dict()
```

Comment:
Why not `self.cell_to_global = {}` (and others). I would say that it's more Pythonic.